### PR TITLE
fix ConditionDirectoryNotEmpty when it comes to a Non-directory file

### LINKED ISSUE #21561: ConditionDirectoryNotEmpty doesn't run as expected when it comes to a Non-directory file
**systemd version the issue has been seen with**
 > systemd-248~rc2

**Used distribution**
 > Fedora release 34 (Thirty Four)

**Linux kernel version used** (`uname -a`)

 > `…`

**CPU architecture issue was seen on**
 > …

**Expected behaviour you didn't see**
 >
The test.service shouldn't be started when /tmp/filename is not a dir.
[root@fedora system]# cat test.service
#/etc/systemd/system/test.service
[Unit]
Description="Test ConditionDirectoryNotEmpty"
ConditionDirectoryNotEmpty=/tmp/filename

**Unexpected behaviour you saw**
 > …
image

**Steps to reproduce the problem**
 > …

**Additional program output to the terminal or log subsystem illustrating the issue**

```text
…
```

### PR DESCRIPTION
**Pull request by jcg190701**
Fixes #21561.

### CODE REVIEW

File: src/shared/condition.c [hunk 1]
```diff
@@ -931,7 +931,7 @@ static int condition_test_directory_not_empty(Condition *c, char **env) {
         assert(c->type == CONDITION_DIRECTORY_NOT_EMPTY);
 
         r = dir_is_empty(c->parameter);
-        return r <= 0 && r != -ENOENT;
+        return r <= 0 && r != -ENOENT && r != -ENOTDIR;
```
Comment (yuwata) [inline src/shared/condition.c]: Please use `IN_SET()` macro:
```
return r <= 0 && !IN_SET(r, -ENOENT, -ENOTDIR);
```

**Review by yuwata — APPROVED** [review]
LGTM. One coding style suggestion.

**Comment by jcg190701** [comment]
Reworked, thanks for your suggestion

File: src/shared/condition.c [hunk 2]
```diff
@@ -898,7 +898,7 @@ static int condition_test_path_is_read_write(Condition *c, char **env) {
         assert(c->parameter);
         assert(c->type == CONDITION_PATH_IS_READ_WRITE);
 
-        return path_is_read_only_fs(c->parameter) <= 0;
+        return path_is_read_only_fs(c->parameter) > 0;
```
Comment (yuwata) [inline src/shared/condition.c]: Hm??? The condition is RW, and the function check RO... Why do you change this??

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/shared/condition.c (hunk 2 above)
Comment (jcg190701) [reply src/shared/condition.c]: oh，I made a mistake. 
But I find that if the path doesn't exist and not null, ConditionPathIsReadWrite will always return true.

Comment (jcg190701) [reply src/shared/condition.c]: in path_is_read_only_fs()

if (statvfs(path, &st) < 0)
       return -errno;

if statvfs() failed， ConditionPathIsReadWrite will always be true.

Comment (jcg190701) [reply src/shared/condition.c]: @yuwata

File: src/shared/condition.c [hunk 3]
```diff
@@ -894,11 +894,15 @@ static int condition_test_path_is_mount_point(Condition *c, char **env) {
...
 static int condition_test_path_is_read_write(Condition *c, char **env) {
-        assert(c);
+        int r;
+
+	assert(c);
         assert(c->parameter);
         assert(c->type == CONDITION_PATH_IS_READ_WRITE);
 
-        return path_is_read_only_fs(c->parameter) <= 0;
+	r = path_is_read_only_fs(c->parameter);
```
Comment (yuwata) [inline src/shared/condition.c]: Nah, spurious indentation...

File: src/shared/condition.c (hunk 2 above)
Comment (yuwata) [reply src/shared/condition.c]: The current logic sounds good for me.

File: src/shared/condition.c [hunk 4]
```diff
@@ -894,11 +894,15 @@ static int condition_test_path_is_mount_point(Condition *c, char **env) {
...
 
 static int condition_test_path_is_read_write(Condition *c, char **env) {
+        int r;
+
         assert(c);
         assert(c->parameter);
         assert(c->type == CONDITION_PATH_IS_READ_WRITE);
 
-        return path_is_read_only_fs(c->parameter) <= 0;
+	r = path_is_read_only_fs(c->parameter);
```
Comment (yuwata) [inline src/shared/condition.c]: Hmm,,, still indentation is strange...

**Review by yuwata — CHANGES_REQUESTED** [review]

**Review by yuwata — APPROVED** [review]
LGTM.

**Pull request merged**